### PR TITLE
Add URI, URL Prisms

### LINKED ISSUE #488: add Uri Prism
```
val uri: Prism[String, java.net.URI]
```

see also https://github.com/julien-truffaut/Monocle/issues/485

### PR DESCRIPTION
**Pull request by n4to4**
Fix #488 , #489

Maybe there are better/simpler ways to generate `Arbitrary`, `Cogen` instances? Any feedback is welcome :)

### CODE REVIEW

File: test/shared/src/test/scala/monocle/TestInstances.scala [hunk 1]
```diff
@@ -196,4 +199,28 @@ trait TestInstances {
...
+      fragment <- Gen.option(idGen)
+    } yield new URI(scheme, ssp, fragment.orNull)
+  }
+
+  implicit def uriCoGen: Cogen[URI] =
+    Cogen[String].contramap[URI](_.toString)
+
+  implicit def urlArbitrary: Arbitrary[URL] = Arbitrary {
+    for {
+      protocol <- Gen.oneOf("http", "https")
```
Comment (xuwei-k) [inline test/shared/src/test/scala/monocle/TestInstances.scala]: > Protocol handlers for the following protocols are guaranteed to exist on the search path :- 
>     http, https, file, and jar

https://docs.oracle.com/javase/8/docs/api/java/net/URL.html#URL-java.lang.String-java.lang.String-int-java.lang.String-

**Comment by n4to4** [comment]
@xuwei-k Thanks for your review.

- I moved `java.net.URL` specific code to the `jvm` specific directory.
  (I'm not sure if this is the best way to separate `js`/`jvm` specific code, though :sweat:)
  
- About adding protocols, it seems that `jar` protocol has its own format: `jar:<url>!/{entry}`,
  and `new URL("jar://host.com:123/path/to/jar")` throws an exception, for example.
  So I added a `Gen[URL]` for `jar`.
  https://docs.oracle.com/javase/8/docs/api/java/net/JarURLConnection.html

**Comment by julien-truffaut** [comment]
LGTM, what do you think @xuwei-k ?

File: test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.net.URL
+import org.scalacheck.{Arbitrary, Cogen, Gen}
+import scalaz.Equal
+
+private [monocle] trait PlatformSpecificTestInstances {
+  implicit val urlEqual = Equal.equalA[URL]
+
+  implicit def urlArbitrary: Arbitrary[URL] = Arbitrary {
+    def idGen = Gen.nonEmptyListOf(Gen.alphaChar).map(_.mkString)
```
Comment (xuwei-k) [inline test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala]: `val idGen`

File: test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      port <- Gen.choose(0, 65535)
+      path <- path("/")
+    } yield new URL(protocol, host, port, path)
+
+    def jar: Gen[URL] = for {
+      url <- url
+      path <- path("!/")
+    } yield new URL("jar:" + url + path)
+
+    Gen.frequency((1, url), (1, jar))
```
Comment (xuwei-k) [inline test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala]: `Gen.oneOf(url, jar)`

or

`Gen.frequency((3, url), (1, jar))`

?

File: test/shared/src/test/scala/monocle/TestInstances.scala [hunk 4]
```diff
@@ -196,4 +198,15 @@ trait TestInstances {
   implicit def uuidCoGen: Cogen[UUID] =
     Cogen[(Long, Long)].contramap[UUID]((u: UUID) => (u.getMostSignificantBits, u.getLeastSignificantBits))
 
+  implicit def uriArbitrary: Arbitrary[URI] = Arbitrary {
+    def idGen = Gen.nonEmptyListOf(Gen.alphaChar).map(_.mkString)
```
Comment (xuwei-k) [inline test/shared/src/test/scala/monocle/TestInstances.scala]: `val idGen`

File: test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  implicit def urlArbitrary: Arbitrary[URL] = Arbitrary {
+    def idGen = Gen.nonEmptyListOf(Gen.alphaChar).map(_.mkString)
+
+    def path(sep: String): Gen[String] = for {
+      length <- Gen.choose(0, 5)
+      path <- Gen.listOfN(length, idGen)
+    } yield path.mkString(sep, "/", "")
+
+    def url: Gen[URL] = for {
```
Comment (xuwei-k) [inline test/jvm/src/test/scala/monocle/PlatformSpecificTestInstances.scala]: `val url`

**Comment by julien-truffaut** [comment]
Great, thank you guys 👍

**Pull request merged**